File: JavaFundamentals/src/MapsLambdaAndStreamAPI/Exercise/LegendaryFarming.java
package MapsLambdaAndStreamAPI.Exercise;

import java.util.*;

public class LegendaryFarming {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        Map<String, Integer> keyMaterials = new HashMap<>();
        Map<String, Integer> junkItems = new LinkedHashMap<>();

        keyMaterials.put("shards", 0);
        keyMaterials.put("fragments", 0);
        keyMaterials.put("motes", 0);

        boolean legendaryObtained = false;
        String legendaryItem = "";

        while (!legendaryObtained) {
            String[] input = scanner.nextLine().toLowerCase().split("\\s+");
            for (int i = 0; i < input.length; i += 2) {
                int quantity = Integer.parseInt(input[i]);
                String material = input[i + 1];

                if (keyMaterials.containsKey(material)) {
                    keyMaterials.put(material, keyMaterials.get(material) + quantity);
                    if (keyMaterials.get(material) >= 250) {
                        legendaryObtained = true;
                        legendaryItem = material;
                        keyMaterials.put(material, keyMaterials.get(material) - 250);
                        break;
                    }
                } else {
                    junkItems.put(material, junkItems.getOrDefault(material, 0) + quantity);
                }
            }
        }

        switch (legendaryItem) {
            case "shards":
                System.out.println("Shadowmourne obtained!");
                break;
            case "fragments":
                System.out.println("Valanyr obtained!");
                break;
            case "motes":
                System.out.println("Dragonwrath obtained!");
                break;
        }

        System.out.println("shards: " + keyMaterials.get("shards"));
        System.out.println("fragments: " + keyMaterials.get("fragments"));
        System.out.println("motes: " + keyMaterials.get("motes"));

        for (Map.Entry<String, Integer> entry : junkItems.entrySet()) {
            System.out.println(entry.getKey() + ": " + entry.getValue());
        }

        scanner.close();
    }
}

File: Fundamentals/Dictionaries - Exercise/legendary_farming.py
materials = input().split(" ")
materials_dict = dict()
is_win_the_race = True
while is_win_the_race:
    for i in range(0, len(materials), 2):
        material = materials[i+1].lower()
        quantity = int(materials[i])
        if material not in materials_dict:
            materials_dict[material] = quantity
        else:
            materials_dict[material] += quantity
        if "shards" in materials_dict.keys():
            if materials_dict["shards"] >= 250:
                print("Shadowmourne obtained!")
                materials_dict["shards"] -= 250
                is_win_the_race = False
                break
        if "fragments" in materials_dict.keys():
            if materials_dict["fragments"] >= 250:
                print("Valanyr obtained!")
                materials_dict["fragments"] -= 250
                is_win_the_race = False
                break
        if 'motes' in materials_dict.keys():
            if materials_dict["motes"] >= 250:
                print("Dragonwrath obtained!")
                materials_dict["motes"] -= 250
                is_win_the_race = False
                break
    if not is_win_the_race:
        break
    materials = input().split(" ")
quantity_of_shards = 0
quantity_of_fragments = 0
quantity_of_motes = 0
if "shards" in materials_dict.keys():
    quantity_of_shards = materials_dict.pop("shards")
if "fragments" in materials_dict.keys():
    quantity_of_fragments = materials_dict.pop("fragments")
if "motes" in materials_dict.keys():
    quantity_of_motes = materials_dict.pop("motes")

print(f"shards: {quantity_of_shards}\nfragments: {quantity_of_fragments}\nmotes: {quantity_of_motes}")
for material, quantity in materials_dict.items():
    print(f"{material}: {quantity}")